src/ophys_etl/decrosstalk/io_utils.py

```diff
@@ -0,0 +1,427 @@
+import h5py
+import json
+import os
+import numpy as np
+
+
+def _write_data_to_h5(file_handle, data_dict):
+    """
+    Utility to write nested dicts to h5py files
+
+    This will result in an hdf5 file in which nested dicts
+    appear as nested directories, i.e.
+
+    data['a'] = True
+    data['b'] = {'c': False, 'd':np.array([1,2,3,4])}
+
+    would result in and hdf5 file with
+
+    data['a'] = True
+    data['b/c'] = False
+    data['b/d'] = np.array([1,2,3,4])
+
+    Parameters
+    ----------
+    file_handle -- a file handle pointing to the file in
+    which we are writing
+
+    data_dict -- the dict of data to be written
+    """
+
+    key_list = list(data_dict.keys())
+    for key in key_list:
+        value = data_dict[key]
+        if isinstance(value, dict):
+            group = file_handle.create_group(str(key))
+            _write_data_to_h5(group, value)
+        else:
+            file_handle.create_dataset(str(key), data=value)
+    return None
+
+
+def write_to_h5(file_name, data_dict, clobber=False):
+    """
+    Write a nested dict to an HDF5 file, treating each level of
+    key as a new directory in the HDF5 file's structure.
+
+    e.g.
+    data_dict['a'] = True
+    data_dict['b'] = {'c': False, 'd':np.array([1,2,3,4])}
+
+    would result in and hdf5 file with
+
+    hdf5_data['a'] = True
+    hdf5_data['b/c'] = False
+    hdf5_data['b/d'] = np.array([1,2,3,4])
+
+
+    Parameters
+    -----------
+    file_name -- the name of the HDF5 file to write
+
+    data_dict -- the dict of data to write
+
+    clobber -- boolean; if False, will not overwrite an existing file
+    (default=False)
+    """
+    if not isinstance(data_dict, dict):
+        msg = '\nInput to write_to_h5 not a dict\n'
+        msg += 'Input was %s' % str(type(data_dict))
+        raise RuntimeError(msg)
+
+    if os.path.exists(file_name) and not clobber:
+        msg = "\n%s\nalready exists; " % file_name
+        msg += "use clobber=True to overwrite"
+        raise RuntimeError(msg)
+    with h5py.File(file_name, mode='w') as file_handle:
+        _write_data_to_h5(file_handle, data_dict)
+    return None
+
+
+def write_basic_json(out_fname, data, clobber=False):
+    """
+    Write data to a json file
+
+    Parameters
+    ----------
+    out_fname -- name of the file to write
+
+    data -- data to be written
+
+    clobber -- boolean; if True, overwrite preexisting
+    out_fname; otherwise, raise an exception if out_fname
+    already exists (default=False)
+    """
+    if not clobber and os.path.exists(out_fname):
+        raise RuntimeError("\n%s\nalready exists" % out_fname)
+    with open(out_fname, 'w') as out_file:
+        out_file.write(json.dumps(data, indent=2, sort_keys=True))
+
+
+# The classes below are implemented to allow us to quickly
```

Comment:
There is now a LIMS strategy for this. Can we just include ouptut formatting for what is expected in production? It seems like if we had to, we/the science team could write translators to the old format.

Reply:
This block of code came about when it became apparent that it would take less work for me to change the output of the module to fit what Natalia's plotting code needed than it would take for Natalia to change her plotting code. Once we have ownership of the plotting code we can think more clearly about what these files should look like. Given that none of this code affects the outputs that LIMS expects (these classes all exist to populate the QC outputs directory), I would like to just leave this as-is and revisit the question of what should be done here in a future ticket.

Is that reasonable?

Reply:
Reasonable